**Financial Data Analysis**

**Problem Definition:** - This dataset contains detailed financial and demographic information of 5,000 individuals, including their income, account balances, transaction history, investment behavior, and loan information. The goal is to analyze and understand individual financial behavior, investment patterns, and risk profiles to support personalized financial decision making. **Dataset source**: Kaggle Datasets

**Data loading and Initial Overview**

1.Installing pandas

In [5]:
pip install pandas


Note: you may need to restart the kernel to use updated packages.


You should consider upgrading via the 'C:\Users\Irene Fathima\AppData\Local\Programs\Python\Python39\python.exe -m pip install --upgrade pip' command.


2.Importing pandas and numpy libraries

In [3]:
import pandas as pd
import numpy as np


In [4]:
df = pd.read_csv('Individual financial profiles.csv')  #this reads a csv file 
df

Age Occupation Risk Tolerance     Investment Goals Income Level  \
0      40     Lawyer           High  Wealth Preservation    $46044.94   
1      30    Teacher            Low  Wealth Preservation    $57169.50   
2      37    Teacher            Low          Speculation    $71760.86   
3      27    Student         Medium          Speculation   $-25488.15   
4      36   Engineer            Low    Income Generation   $106777.95   
...   ...        ...            ...                  ...          ...   
4995   21   Engineer           High    Income Generation    $92640.59   
4996   20    Student            Low  Wealth Preservation   $100107.46   
4997   53     Lawyer            Low          Speculation     $6058.18   
4998   75    Student         Medium               Growth    $76214.25   
4999   33    Retiree            Low               Growth    $62638.63   

                                                Address Account Balance  \
0           7168 Moody Meadow\nHernandezshire, PW 06016       $44653.26   
1          50001 Hector Square\nWest Luisfurt, MA 51935       $29175.47   
2           997 James Isle\nNorth Rebeccafurt, RI 13366       $86141.59   
3      1607 Joshua Camp Apt. 634\nConleymouth, CT 66479        $1000.00   
4           96690 Campbell Common\nDwaynetown, MI 71857       $77918.97   
...                                                 ...             ...   
4995          43377 Brooke Gateway\nKyleville, IL 60381      $100000.00   
4996  1156 Christopher Mountain Suite 011\nEast Jimm...       $29879.73   
4997  92818 Mary Garden Suite 343\nGuerramouth, DE 6...        $9673.78   
4998  6197 Stephanie Circles Apt. 765\nCherylstad, N...       $90780.00   
4999                   Unit 4064 Box 6179\nDPO AA 30544       $62218.05   

       Deposits Withdrawals  Transfers International Transfers Investments  \
0      $9156.01    $9327.70   $3647.92                  $82.80    $8729.30   
1      $5933.22    $8671.60   $6729.86                 $819.69    $4545.18   
2     $22583.11   $16468.58   $6032.53                $1526.33    $8251.45   
3       $299.47     $289.09    $109.83                   $3.51     $195.73   
4     $13853.25   $22699.32  $12189.32                $2152.51    $9725.21   
...         ...         ...        ...                     ...         ...   
4995  $24890.81   $17829.06  $21150.16                $5161.83   $32126.21   
4996   $5384.31    $6262.80   $3931.47                 $757.33    $2529.55   
4997   $3274.37    $1832.16   $1742.00                 $354.33     $898.41   
4998  $35679.58   $12022.55  $12832.11                $3389.92    $6370.88   
4999  $22903.01   $16866.44   $4295.50                 $484.80    $3886.95   

     Loan Amount      Loan Purpose Employment Status  Loan Term (Months)  \
0      $27010.93  Medical Expenses           Retired                  36   
1      $31266.97     Auto Purchase           Retired                  36   
2      $41260.58     Auto Purchase          Employed                  12   
3       $5000.00    Small Business           Retired                  60   
4      $50000.00  Medical Expenses          Employed                  24   
...          ...               ...               ...                 ...   
4995   $50000.00  Medical Expenses        Unemployed                  24   
4996   $50000.00     Auto Purchase     Self-Employed                  24   
4997    $7054.49  Medical Expenses     Self-Employed                  60   
4998   $34847.26  Home Improvement           Retired                  48   
4999   $12459.69  Medical Expenses     Self-Employed                  12   

     Interest Rate Loan Status  \
0           11.94%     pending   
1            8.08%    approved   
2           13.07%     pending   
3            6.73%    approved   
4           10.60%    approved   
...            ...         ...   
4995        11.41%    approved   
4996         9.38%    approved   
4997        10.02%     pending   
4998         8.49%    approved   
4999

4.To find Number of rows and columns in the dataset.

In [77]:
df.shape

(5000, 19)

There are 5000 rows and 19 columns in the dataset

5.To find datatypes of each column

In [7]:
df.dtypes

Age                         int64
Occupation                 object
Risk Tolerance             object
Investment Goals           object
Income Level               object
Address                    object
Account Balance            object
Deposits                   object
Withdrawals                object
Transfers                  object
International Transfers    object
Investments                object
Loan Amount                object
Loan Purpose               object
Employment Status          object
Loan Term (Months)          int64
Interest Rate              object
Loan Status                object
Transaction Description    object
dtype: object

int64 - integer numbers , object - represents text or mixed numbers(e.g.,strings)

6.To view your data to printout a few rows to keep a visual reference

In [8]:
df.head()

Age Occupation Risk Tolerance     Investment Goals Income Level  \
0   40     Lawyer           High  Wealth Preservation    $46044.94   
1   30    Teacher            Low  Wealth Preservation    $57169.50   
2   37    Teacher            Low          Speculation    $71760.86   
3   27    Student         Medium          Speculation   $-25488.15   
4   36   Engineer            Low    Income Generation   $106777.95   

                                            Address Account Balance  \
0       7168 Moody Meadow\nHernandezshire, PW 06016       $44653.26   
1      50001 Hector Square\nWest Luisfurt, MA 51935       $29175.47   
2       997 James Isle\nNorth Rebeccafurt, RI 13366       $86141.59   
3  1607 Joshua Camp Apt. 634\nConleymouth, CT 66479        $1000.00   
4       96690 Campbell Common\nDwaynetown, MI 71857       $77918.97   

    Deposits Withdrawals  Transfers International Transfers Investments  \
0   $9156.01    $9327.70   $3647.92                  $82.80    $8729.30   
1   $5933.22    $8671.60   $6729.86                 $819.69    $4545.18   
2  $22583.11   $16468.58   $6032.53                $1526.33    $8251.45   
3    $299.47     $289.09    $109.83                   $3.51     $195.73   
4  $13853.25   $22699.32  $12189.32                $2152.51    $9725.21   

  Loan Amount      Loan Purpose Employment Status  Loan Term (Months)  \
0   $27010.93  Medical Expenses           Retired                  36   
1   $31266.97     Auto Purchase           Retired                  36   
2   $41260.58     Auto Purchase          Employed                  12   
3    $5000.00    Small Business           Retired                  60   
4   $50000.00  Medical Expenses          Employed                  24   

  Interest Rate Loan Status                            Transaction Description  
0        11.94%     pending  Electronics transaction of $706.18 at Sanders,...  
1         8.08%    approved            Transaction at Evans-Smith for $2250.03  
2        13.07%     pending  Purchase at Taylor-Gutierrez for $615.85 on 20...  
3         6.73%    approved    Travel transaction of $4852.39 at Jones-Russell  
4        10.60%    approved  Purchase at Jensen, Porter and Webb for $562.2...

This provides details of the first five rows

7.To see the details of last five rows, we use


In [9]:
df.tail()

Age Occupation Risk Tolerance     Investment Goals Income Level  \
4995   21   Engineer           High    Income Generation    $92640.59   
4996   20    Student            Low  Wealth Preservation   $100107.46   
4997   53     Lawyer            Low          Speculation     $6058.18   
4998   75    Student         Medium               Growth    $76214.25   
4999   33    Retiree            Low               Growth    $62638.63   

                                                Address Account Balance  \
4995          43377 Brooke Gateway\nKyleville, IL 60381      $100000.00   
4996  1156 Christopher Mountain Suite 011\nEast Jimm...       $29879.73   
4997  92818 Mary Garden Suite 343\nGuerramouth, DE 6...        $9673.78   
4998  6197 Stephanie Circles Apt. 765\nCherylstad, N...       $90780.00   
4999                   Unit 4064 Box 6179\nDPO AA 30544       $62218.05   

       Deposits Withdrawals  Transfers International Transfers Investments  \
4995  $24890.81   $17829.06  $21150.16                $5161.83   $32126.21   
4996   $5384.31    $6262.80   $3931.47                 $757.33    $2529.55   
4997   $3274.37    $1832.16   $1742.00                 $354.33     $898.41   
4998  $35679.58   $12022.55  $12832.11                $3389.92    $6370.88   
4999  $22903.01   $16866.44   $4295.50                 $484.80    $3886.95   

     Loan Amount      Loan Purpose Employment Status  Loan Term (Months)  \
4995   $50000.00  Medical Expenses        Unemployed                  24   
4996   $50000.00     Auto Purchase     Self-Employed                  24   
4997    $7054.49  Medical Expenses     Self-Employed                  60   
4998   $34847.26  Home Improvement           Retired                  48   
4999   $12459.69  Medical Expenses     Self-Employed                  12   

     Interest Rate Loan Status  \
4995        11.41%    approved   
4996         9.38%    approved   
4997        10.02%     pending   
4998         8.49%    approved   
4999         7.95%    rejected   

                                Transaction Description  
4995  Payment of $1284.07 at Curtis, Perez and Sandoval  
4996  $2881.42 spent at Smith, Martin and Velez on 2...  
4997  Transaction at Brown, Rubio and Nolan for $324...  
4998  Transaction at Tucker, Tyler and Jacobson for ...  
4999                 Payment of $3027.33 at Howard-Case

details of last 5 rows

8.To get the details regarding data

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 19 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   Age                      5000 non-null   int64 
 1   Occupation               5000 non-null   object
 2   Risk Tolerance           5000 non-null   object
 3   Investment Goals         5000 non-null   object
 4   Income Level             5000 non-null   object
 5   Address                  5000 non-null   object
 6   Account Balance          5000 non-null   object
 7   Deposits                 5000 non-null   object
 8   Withdrawals              5000 non-null   object
 9   Transfers                5000 non-null   object
 10  International Transfers  5000 non-null   object
 11  Investments              5000 non-null   object
 12  Loan Amount              5000 non-null   object
 13  Loan Purpose             5000 non-null   object
 14  Employment Status        5000 non-null  

This shows the information regarding data, such as number of rows and columns in the dataset, null values present in the dataset, datatypes of each column, memory usage of the dataset.Here shows no null values in the dataset

9.describe() function shows the statistical summary of numerical data

In [11]:
df.describe()

Age  Loan Term (Months)
count  5000.000000         5000.000000
mean     51.875600           36.235200
std      19.579888           17.014705
min      18.000000           12.000000
25%      35.000000           24.000000
50%      52.000000           36.000000
75%      69.000000           48.000000
max      85.000000           60.000000

It shows the numerical columns with continuous data

Age distribution: avg age is 52 years.age ranges from 18 to 85 years.Standard deviation is 19.6 suggests a fairly wide age distribution.
Loan Term Distribution:Average loan term is 36 months. Loan Terms range from 12 to 60 months.

10.For column names


In [12]:
df.columns

Index(['Age', 'Occupation', 'Risk Tolerance', 'Investment Goals',
       'Income Level', 'Address', 'Account Balance', 'Deposits', 'Withdrawals',
       'Transfers', 'International Transfers', 'Investments', 'Loan Amount',
       'Loan Purpose', 'Employment Status', 'Loan Term (Months)',
       'Interest Rate', 'Loan Status', 'Transaction Description'],
      dtype='object')

**Handling Missing Values**

To check for missing values

In [13]:
df.isnull()

Age  Occupation  Risk Tolerance  Investment Goals  Income Level  \
0     False       False           False             False         False   
1     False       False           False             False         False   
2     False       False           False             False         False   
3     False       False           False             False         False   
4     False       False           False             False         False   
...     ...         ...             ...               ...           ...   
4995  False       False           False             False         False   
4996  False       False           False             False         False   
4997  False       False           False             False         False   
4998  False       False           False             False         False   
4999  False       False           False             False         False   

      Address  Account Balance  Deposits  Withdrawals  Transfers  \
0       False            False     False        False      False   
1       False            False     False        False      False   
2       False            False     False        False      False   
3       False            False     False        False      False   
4       False            False     False        False      False   
...       ...              ...       ...          ...        ...   
4995    False            False     False        False      False   
4996    False            False     False        False      False   
4997    False            False     False        False      False   
4998    False            False     False        False      False   
4999    False            False     False        False      False   

      International Transfers  Investments  Loan Amount  Loan Purpose  \
0                       False        False        False         False   
1                       False        False        False         False   
2                       False        False        False         False   
3                       False        False        False         False   
4                       False        False        False         False   
...                       ...          ...          ...           ...   
4995                    False        False        False         False   
4996                    False        False        False         False   
4997                    False        False        False         False   
4998                    False        False        False         False   
4999                    False        False        False         False   

      Employment Status  Loan Term (Months)  Interest Rate  Loan Status  \
0                 False               False          False        False   
1                 False               False          False        False   
2                 False               False          False        False   
3                 False               False          False        False   
4                 False               False          False        False   
...                 ...                 ...            ...          ...   
4995              False               False          False        False   
4996              False               False          False        False   
4997              False               False          False        False   
4998              False               False          False        False   
4999              False               False          False        False   

      Transaction Description  
0                       False  
1                       False  
2                       False  
3                       False  
4                       False  
...                       ...  
4995                    False  
4996                    False  
4997                    False  
4998                    False  
4999                    False  

[5000 rows x 19 columns]

In [14]:
df.isnull().sum()

Age                        0
Occupation                 0
Risk Tolerance             0
Investment Goals           0
Income Level               0
Address                    0
Account Balance            0
Deposits                   0
Withdrawals                0
Transfers                  0
International Transfers    0
Investments                0
Loan Amount                0
Loan Purpose               0
Employment Status          0
Loan Term (Months)         0
Interest Rate              0
Loan Status                0
Transaction Description    0
dtype: int64

This shows the number of null values in each column. Here shows no null values so given as zero.

In [15]:
df.isnull().sum().sum()


np.int64(0)

This shows total number of null values in the dataset.

no null values present in the dataset

**Removing Duplicates**

Checking for duplicate rows

In [16]:
df.duplicated().sum()

np.int64(0)

To view duplicate rows

In [17]:
df[df.duplicated()]


Empty DataFrame
Columns: [Age, Occupation, Risk Tolerance, Investment Goals, Income Level, Address, Account Balance, Deposits, Withdrawals, Transfers, International Transfers, Investments, Loan Amount, Loan Purpose, Employment Status, Loan Term (Months), Interest Rate, Loan Status, Transaction Description]
Index: []

In [21]:
# no duplicate rows

In [18]:
df.drop_duplicates(subset=None,keep='first',inplace=False)  


Age Occupation Risk Tolerance     Investment Goals Income Level  \
0      40     Lawyer           High  Wealth Preservation    $46044.94   
1      30    Teacher            Low  Wealth Preservation    $57169.50   
2      37    Teacher            Low          Speculation    $71760.86   
3      27    Student         Medium          Speculation   $-25488.15   
4      36   Engineer            Low    Income Generation   $106777.95   
...   ...        ...            ...                  ...          ...   
4995   21   Engineer           High    Income Generation    $92640.59   
4996   20    Student            Low  Wealth Preservation   $100107.46   
4997   53     Lawyer            Low          Speculation     $6058.18   
4998   75    Student         Medium               Growth    $76214.25   
4999   33    Retiree            Low               Growth    $62638.63   

                                                Address Account Balance  \
0           7168 Moody Meadow\nHernandezshire, PW 06016       $44653.26   
1          50001 Hector Square\nWest Luisfurt, MA 51935       $29175.47   
2           997 James Isle\nNorth Rebeccafurt, RI 13366       $86141.59   
3      1607 Joshua Camp Apt. 634\nConleymouth, CT 66479        $1000.00   
4           96690 Campbell Common\nDwaynetown, MI 71857       $77918.97   
...                                                 ...             ...   
4995          43377 Brooke Gateway\nKyleville, IL 60381      $100000.00   
4996  1156 Christopher Mountain Suite 011\nEast Jimm...       $29879.73   
4997  92818 Mary Garden Suite 343\nGuerramouth, DE 6...        $9673.78   
4998  6197 Stephanie Circles Apt. 765\nCherylstad, N...       $90780.00   
4999                   Unit 4064 Box 6179\nDPO AA 30544       $62218.05   

       Deposits Withdrawals  Transfers International Transfers Investments  \
0      $9156.01    $9327.70   $3647.92                  $82.80    $8729.30   
1      $5933.22    $8671.60   $6729.86                 $819.69    $4545.18   
2     $22583.11   $16468.58   $6032.53                $1526.33    $8251.45   
3       $299.47     $289.09    $109.83                   $3.51     $195.73   
4     $13853.25   $22699.32  $12189.32                $2152.51    $9725.21   
...         ...         ...        ...                     ...         ...   
4995  $24890.81   $17829.06  $21150.16                $5161.83   $32126.21   
4996   $5384.31    $6262.80   $3931.47                 $757.33    $2529.55   
4997   $3274.37    $1832.16   $1742.00                 $354.33     $898.41   
4998  $35679.58   $12022.55  $12832.11                $3389.92    $6370.88   
4999  $22903.01   $16866.44   $4295.50                 $484.80    $3886.95   

     Loan Amount      Loan Purpose Employment Status  Loan Term (Months)  \
0      $27010.93  Medical Expenses           Retired                  36   
1      $31266.97     Auto Purchase           Retired                  36   
2      $41260.58     Auto Purchase          Employed                  12   
3       $5000.00    Small Business           Retired                  60   
4      $50000.00  Medical Expenses          Employed                  24   
...          ...               ...               ...                 ...   
4995   $50000.00  Medical Expenses        Unemployed                  24   
4996   $50000.00     Auto Purchase     Self-Employed                  24   
4997    $7054.49  Medical Expenses     Self-Employed                  60   
4998   $34847.26  Home Improvement           Retired                  48   
4999   $12459.69  Medical Expenses     Self-Employed                  12   

     Interest Rate Loan Status  \
0           11.94%     pending   
1            8.08%    approved   
2           13.07%     pending   
3            6.73%    approved   
4           10.60%    approved   
...            ...         ...   
4995        11.41%    approved   
4996         9.38%    approved   
4997        10.02%     pending   
4998         8.49%    approved   
4999

no duplicates found

**Correcting Datatypes**

In [19]:
df['Account Balance'] = pd.to_numeric(df['Account Balance'].replace('[\$,]', '', regex=True))
df


Age Occupation Risk Tolerance     Investment Goals Income Level  \
0      40     Lawyer           High  Wealth Preservation    $46044.94   
1      30    Teacher            Low  Wealth Preservation    $57169.50   
2      37    Teacher            Low          Speculation    $71760.86   
3      27    Student         Medium          Speculation   $-25488.15   
4      36   Engineer            Low    Income Generation   $106777.95   
...   ...        ...            ...                  ...          ...   
4995   21   Engineer           High    Income Generation    $92640.59   
4996   20    Student            Low  Wealth Preservation   $100107.46   
4997   53     Lawyer            Low          Speculation     $6058.18   
4998   75    Student         Medium               Growth    $76214.25   
4999   33    Retiree            Low               Growth    $62638.63   

                                                Address  Account Balance  \
0           7168 Moody Meadow\nHernandezshire, PW 06016         44653.26   
1          50001 Hector Square\nWest Luisfurt, MA 51935         29175.47   
2           997 James Isle\nNorth Rebeccafurt, RI 13366         86141.59   
3      1607 Joshua Camp Apt. 634\nConleymouth, CT 66479          1000.00   
4           96690 Campbell Common\nDwaynetown, MI 71857         77918.97   
...                                                 ...              ...   
4995          43377 Brooke Gateway\nKyleville, IL 60381        100000.00   
4996  1156 Christopher Mountain Suite 011\nEast Jimm...         29879.73   
4997  92818 Mary Garden Suite 343\nGuerramouth, DE 6...          9673.78   
4998  6197 Stephanie Circles Apt. 765\nCherylstad, N...         90780.00   
4999                   Unit 4064 Box 6179\nDPO AA 30544         62218.05   

       Deposits Withdrawals  Transfers International Transfers Investments  \
0      $9156.01    $9327.70   $3647.92                  $82.80    $8729.30   
1      $5933.22    $8671.60   $6729.86                 $819.69    $4545.18   
2     $22583.11   $16468.58   $6032.53                $1526.33    $8251.45   
3       $299.47     $289.09    $109.83                   $3.51     $195.73   
4     $13853.25   $22699.32  $12189.32                $2152.51    $9725.21   
...         ...         ...        ...                     ...         ...   
4995  $24890.81   $17829.06  $21150.16                $5161.83   $32126.21   
4996   $5384.31    $6262.80   $3931.47                 $757.33    $2529.55   
4997   $3274.37    $1832.16   $1742.00                 $354.33     $898.41   
4998  $35679.58   $12022.55  $12832.11                $3389.92    $6370.88   
4999  $22903.01   $16866.44   $4295.50                 $484.80    $3886.95   

     Loan Amount      Loan Purpose Employment Status  Loan Term (Months)  \
0      $27010.93  Medical Expenses           Retired                  36   
1      $31266.97     Auto Purchase           Retired                  36   
2      $41260.58     Auto Purchase          Employed                  12   
3       $5000.00    Small Business           Retired                  60   
4      $50000.00  Medical Expenses          Employed                  24   
...          ...               ...               ...                 ...   
4995   $50000.00  Medical Expenses        Unemployed                  24   
4996   $50000.00     Auto Purchase     Self-Employed                  24   
4997    $7054.49  Medical Expenses     Self-Employed                  60   
4998   $34847.26  Home Improvement           Retired                  48   
4999   $12459.69  Medical Expenses     Self-Employed                  12   

     Interest Rate Loan Status  \
0           11.94%     pending   
1            8.08%    approved   
2           13.07%     pending   
3            6.73%    approved   
4           10.60%    approved   
...            ...         ...   
4995        11.41%    approved   
4996         9.38%    approved   
4997        10.02%     pending   
4998         8.49%    appr

In [21]:
df['Withdrawals'] = pd.to_numeric(df['Withdrawals'].replace('[\$,]', '', regex=True))
df['Deposits'] = pd.to_numeric(df['Deposits'].replace('[\$,]', '', regex=True))
df['Transfers'] = pd.to_numeric(df['Transfers'].replace('[\$,]', '', regex=True))
df['International Transfers'] = pd.to_numeric(df['International Transfers'].replace('[\$,]', '', regex=True))
df['Loan Amount'] = pd.to_numeric(df['Loan Amount'].replace('[\$,]', '', regex=True))
df['Investments'] = pd.to_numeric(df['Investments'].replace('[\$,]', '', regex=True))




In [24]:
df


Age Occupation Risk Tolerance     Investment Goals Income Level  \
0      40     Lawyer           High  Wealth Preservation    $46044.94   
1      30    Teacher            Low  Wealth Preservation    $57169.50   
2      37    Teacher            Low          Speculation    $71760.86   
3      27    Student         Medium          Speculation   $-25488.15   
4      36   Engineer            Low    Income Generation   $106777.95   
...   ...        ...            ...                  ...          ...   
4995   21   Engineer           High    Income Generation    $92640.59   
4996   20    Student            Low  Wealth Preservation   $100107.46   
4997   53     Lawyer            Low          Speculation     $6058.18   
4998   75    Student         Medium               Growth    $76214.25   
4999   33    Retiree            Low               Growth    $62638.63   

                                                Address  Account Balance  \
0           7168 Moody Meadow\nHernandezshire, PW 06016         44653.26   
1          50001 Hector Square\nWest Luisfurt, MA 51935         29175.47   
2           997 James Isle\nNorth Rebeccafurt, RI 13366         86141.59   
3      1607 Joshua Camp Apt. 634\nConleymouth, CT 66479          1000.00   
4           96690 Campbell Common\nDwaynetown, MI 71857         77918.97   
...                                                 ...              ...   
4995          43377 Brooke Gateway\nKyleville, IL 60381        100000.00   
4996  1156 Christopher Mountain Suite 011\nEast Jimm...         29879.73   
4997  92818 Mary Garden Suite 343\nGuerramouth, DE 6...          9673.78   
4998  6197 Stephanie Circles Apt. 765\nCherylstad, N...         90780.00   
4999                   Unit 4064 Box 6179\nDPO AA 30544         62218.05   

      Deposits  Withdrawals  Transfers  International Transfers  Investments  \
0      9156.01      9327.70    3647.92                    82.80      8729.30   
1      5933.22      8671.60    6729.86                   819.69      4545.18   
2     22583.11     16468.58    6032.53                  1526.33      8251.45   
3       299.47       289.09     109.83                     3.51       195.73   
4     13853.25     22699.32   12189.32                  2152.51      9725.21   
...        ...          ...        ...                      ...          ...   
4995  24890.81     17829.06   21150.16                  5161.83     32126.21   
4996   5384.31      6262.80    3931.47                   757.33      2529.55   
4997   3274.37      1832.16    1742.00                   354.33       898.41   
4998  35679.58     12022.55   12832.11                  3389.92      6370.88   
4999  22903.01     16866.44    4295.50                   484.80      3886.95   

      Loan Amount      Loan Purpose Employment Status  Loan Term (Months)  \
0        27010.93  Medical Expenses           Retired                  36   
1        31266.97     Auto Purchase           Retired                  36   
2        41260.58     Auto Purchase          Employed                  12   
3         5000.00    Small Business           Retired                  60   
4        50000.00  Medical Expenses          Employed                  24   
...           ...               ...               ...                 ...   
4995     50000.00  Medical Expenses        Unemployed                  24   
4996     50000.00     Auto Purchase     Self-Employed                  24   
4997      7054.49  Medical Expenses     Self-Employed                  60   
4998     34847.26  Home Improvement           Retired                  48   
4999     12459.69  Medical Expenses     Self-Employed                  12   

     Interest Rate Loan Status  \
0           11.94%     pending   
1            8.08%    approved   
2           13.07%     pending   
3            6.73%    approved   
4           10.60%    approved   
...            ...         ...   
4995        11.41%    approved   
4996         9.38%    approved   
4997        10.02%     p

In [22]:
df['Income Level'] = pd.to_numeric(df['Income Level'].replace('[\$,]', '', regex=True))



In [23]:
df['Interest Rate'] = pd.to_numeric(df['Interest Rate'].str.replace('%', ''))


Numerical columns stored as object which likely contain numbers stored as text or strings is converted

In [25]:
df['Risk Tolerance'] = df['Risk Tolerance'].astype('category')
df['Occupation'] = df['Occupation'].astype('category')

In [26]:
df['Investment Goals'] = df['Investment Goals'].astype('category')
df['Loan Status'] = df['Loan Status'].astype('category')

In [27]:
df['Loan Purpose'] = df['Loan Purpose'].astype('category')
df['Employment Status'] = df['Employment Status'].astype('category')
df['Transaction Description'] = df['Transaction Description'].astype('category')



In [86]:
df


Age Occupation Risk Tolerance     Investment Goals Income Level  \
0      40     Lawyer           High  Wealth Preservation    $46044.94   
1      30    Teacher            Low  Wealth Preservation    $57169.50   
2      37    Teacher            Low          Speculation    $71760.86   
3      27    Student         Medium          Speculation   $-25488.15   
4      36   Engineer            Low    Income Generation   $106777.95   
...   ...        ...            ...                  ...          ...   
4995   21   Engineer           High    Income Generation    $92640.59   
4996   20    Student            Low  Wealth Preservation   $100107.46   
4997   53     Lawyer            Low          Speculation     $6058.18   
4998   75    Student         Medium               Growth    $76214.25   
4999   33    Retiree            Low               Growth    $62638.63   

                                                Address Account Balance  \
0           7168 Moody Meadow\nHernandezshire, PW 06016       $44653.26   
1          50001 Hector Square\nWest Luisfurt, MA 51935       $29175.47   
2           997 James Isle\nNorth Rebeccafurt, RI 13366       $86141.59   
3      1607 Joshua Camp Apt. 634\nConleymouth, CT 66479        $1000.00   
4           96690 Campbell Common\nDwaynetown, MI 71857       $77918.97   
...                                                 ...             ...   
4995          43377 Brooke Gateway\nKyleville, IL 60381      $100000.00   
4996  1156 Christopher Mountain Suite 011\nEast Jimm...       $29879.73   
4997  92818 Mary Garden Suite 343\nGuerramouth, DE 6...        $9673.78   
4998  6197 Stephanie Circles Apt. 765\nCherylstad, N...       $90780.00   
4999                   Unit 4064 Box 6179\nDPO AA 30544       $62218.05   

       Deposits Withdrawals  Transfers International Transfers Investments  \
0      $9156.01    $9327.70   $3647.92                  $82.80    $8729.30   
1      $5933.22    $8671.60   $6729.86                 $819.69    $4545.18   
2     $22583.11   $16468.58   $6032.53                $1526.33    $8251.45   
3       $299.47     $289.09    $109.83                   $3.51     $195.73   
4     $13853.25   $22699.32  $12189.32                $2152.51    $9725.21   
...         ...         ...        ...                     ...         ...   
4995  $24890.81   $17829.06  $21150.16                $5161.83   $32126.21   
4996   $5384.31    $6262.80   $3931.47                 $757.33    $2529.55   
4997   $3274.37    $1832.16   $1742.00                 $354.33     $898.41   
4998  $35679.58   $12022.55  $12832.11                $3389.92    $6370.88   
4999  $22903.01   $16866.44   $4295.50                 $484.80    $3886.95   

     Loan Amount      Loan Purpose Employment Status  Loan Term (Months)  \
0      $27010.93  Medical Expenses           Retired                  36   
1      $31266.97     Auto Purchase           Retired                  36   
2      $41260.58     Auto Purchase          Employed                  12   
3       $5000.00    Small Business           Retired                  60   
4      $50000.00  Medical Expenses          Employed                  24   
...          ...               ...               ...                 ...   
4995   $50000.00  Medical Expenses        Unemployed                  24   
4996   $50000.00     Auto Purchase     Self-Employed                  24   
4997    $7054.49  Medical Expenses     Self-Employed                  60   
4998   $34847.26  Home Improvement           Retired                  48   
4999   $12459.69  Medical Expenses     Self-Employed                  12   

      Interest Rate Loan Status  \
0             11.94     pending   
1              8.08    approved   
2             13.07     pending   
3              6.73    approved   
4             10.60    approved   
...             ...         ...   
4995          11.41    approved   
4996           9.38    approved   
4997          10.02     pending   
4998           8.49    appro

Text columns is converted into category type.

**Creating derived columns**

1.Net flow/Cash flow for each individual can be derived from the data.Cash flow represents the movement of money in an out of an individuals account over a period.

In [28]:
df['Cash Flow'] = df['Deposits'] - (df['Withdrawals'] + df['Transfers'] + df['International Transfers'])
df

Age Occupation Risk Tolerance     Investment Goals  Income Level  \
0      40     Lawyer           High  Wealth Preservation      46044.94   
1      30    Teacher            Low  Wealth Preservation      57169.50   
2      37    Teacher            Low          Speculation      71760.86   
3      27    Student         Medium          Speculation     -25488.15   
4      36   Engineer            Low    Income Generation     106777.95   
...   ...        ...            ...                  ...           ...   
4995   21   Engineer           High    Income Generation      92640.59   
4996   20    Student            Low  Wealth Preservation     100107.46   
4997   53     Lawyer            Low          Speculation       6058.18   
4998   75    Student         Medium               Growth      76214.25   
4999   33    Retiree            Low               Growth      62638.63   

                                                Address  Account Balance  \
0           7168 Moody Meadow\nHernandezshire, PW 06016         44653.26   
1          50001 Hector Square\nWest Luisfurt, MA 51935         29175.47   
2           997 James Isle\nNorth Rebeccafurt, RI 13366         86141.59   
3      1607 Joshua Camp Apt. 634\nConleymouth, CT 66479          1000.00   
4           96690 Campbell Common\nDwaynetown, MI 71857         77918.97   
...                                                 ...              ...   
4995          43377 Brooke Gateway\nKyleville, IL 60381        100000.00   
4996  1156 Christopher Mountain Suite 011\nEast Jimm...         29879.73   
4997  92818 Mary Garden Suite 343\nGuerramouth, DE 6...          9673.78   
4998  6197 Stephanie Circles Apt. 765\nCherylstad, N...         90780.00   
4999                   Unit 4064 Box 6179\nDPO AA 30544         62218.05   

      Deposits  Withdrawals  Transfers  International Transfers  Investments  \
0      9156.01      9327.70    3647.92                    82.80      8729.30   
1      5933.22      8671.60    6729.86                   819.69      4545.18   
2     22583.11     16468.58    6032.53                  1526.33      8251.45   
3       299.47       289.09     109.83                     3.51       195.73   
4     13853.25     22699.32   12189.32                  2152.51      9725.21   
...        ...          ...        ...                      ...          ...   
4995  24890.81     17829.06   21150.16                  5161.83     32126.21   
4996   5384.31      6262.80    3931.47                   757.33      2529.55   
4997   3274.37      1832.16    1742.00                   354.33       898.41   
4998  35679.58     12022.55   12832.11                  3389.92      6370.88   
4999  22903.01     16866.44    4295.50                   484.80      3886.95   

      Loan Amount      Loan Purpose Employment Status  Loan Term (Months)  \
0        27010.93  Medical Expenses           Retired                  36   
1        31266.97     Auto Purchase           Retired                  36   
2        41260.58     Auto Purchase          Employed                  12   
3         5000.00    Small Business           Retired                  60   
4        50000.00  Medical Expenses          Employed                  24   
...           ...               ...               ...                 ...   
4995     50000.00  Medical Expenses        Unemployed                  24   
4996     50000.00     Auto Purchase     Self-Employed                  24   
4997      7054.49  Medical Expenses     Self-Employed                  60   
4998     34847.26  Home Improvement           Retired                  48   
4999     12459.69  Medical Expenses     Self-Employed                  12   

      Interest Rate Loan Status  \
0             11.94     pending   
1              8.08    approved   
2             13.07     pending   
3              6.73    approved   
4             10.60    approved   
...             ...         ...   
4995          11.41    approved   
4996           9.38    approved   
499

Positive value of cashflow - implies more money coming in than going out or the person is saving money.
Negative value of cashflow - implies more money is going out or the person is accumulating debt.
Zero - implies a balanced cashflow. money coming in equals money going out.

2.Loan to Income ratio

In [29]:
df['Loan to Income Ratio'] = df['Loan Amount'] / df['Account Balance']
df



Age Occupation Risk Tolerance     Investment Goals  Income Level  \
0      40     Lawyer           High  Wealth Preservation      46044.94   
1      30    Teacher            Low  Wealth Preservation      57169.50   
2      37    Teacher            Low          Speculation      71760.86   
3      27    Student         Medium          Speculation     -25488.15   
4      36   Engineer            Low    Income Generation     106777.95   
...   ...        ...            ...                  ...           ...   
4995   21   Engineer           High    Income Generation      92640.59   
4996   20    Student            Low  Wealth Preservation     100107.46   
4997   53     Lawyer            Low          Speculation       6058.18   
4998   75    Student         Medium               Growth      76214.25   
4999   33    Retiree            Low               Growth      62638.63   

                                                Address  Account Balance  \
0           7168 Moody Meadow\nHernandezshire, PW 06016         44653.26   
1          50001 Hector Square\nWest Luisfurt, MA 51935         29175.47   
2           997 James Isle\nNorth Rebeccafurt, RI 13366         86141.59   
3      1607 Joshua Camp Apt. 634\nConleymouth, CT 66479          1000.00   
4           96690 Campbell Common\nDwaynetown, MI 71857         77918.97   
...                                                 ...              ...   
4995          43377 Brooke Gateway\nKyleville, IL 60381        100000.00   
4996  1156 Christopher Mountain Suite 011\nEast Jimm...         29879.73   
4997  92818 Mary Garden Suite 343\nGuerramouth, DE 6...          9673.78   
4998  6197 Stephanie Circles Apt. 765\nCherylstad, N...         90780.00   
4999                   Unit 4064 Box 6179\nDPO AA 30544         62218.05   

      Deposits  Withdrawals  Transfers  ...  Investments  Loan Amount  \
0      9156.01      9327.70    3647.92  ...      8729.30     27010.93   
1      5933.22      8671.60    6729.86  ...      4545.18     31266.97   
2     22583.11     16468.58    6032.53  ...      8251.45     41260.58   
3       299.47       289.09     109.83  ...       195.73      5000.00   
4     13853.25     22699.32   12189.32  ...      9725.21     50000.00   
...        ...          ...        ...  ...          ...          ...   
4995  24890.81     17829.06   21150.16  ...     32126.21     50000.00   
4996   5384.31      6262.80    3931.47  ...      2529.55     50000.00   
4997   3274.37      1832.16    1742.00  ...       898.41      7054.49   
4998  35679.58     12022.55   12832.11  ...      6370.88     34847.26   
4999  22903.01     16866.44    4295.50  ...      3886.95     12459.69   

          Loan Purpose Employment Status Loan Term (Months)  Interest Rate  \
0     Medical Expenses           Retired                 36          11.94   
1        Auto Purchase           Retired                 36           8.08   
2        Auto Purchase          Employed                 12          13.07   
3       Small Business           Retired                 60           6.73   
4     Medical Expenses          Employed                 24          10.60   
...                ...               ...                ...            ...   
4995  Medical Expenses        Unemployed                 24          11.41   
4996     Auto Purchase     Self-Employed                 24           9.38   
4997  Medical Expenses     Self-Employed                 60          10.02   
4998  Home Improvement           Retired                 48           8.49   
4999  Medical Expenses     Self-Employed                 12           7.95   

      Loan Status                            Transaction Description  \
0         pending  Electronics transaction of $706.18 at Sanders,...   
1        approved            Transaction at Evans-Smith for $2250.03   
2         pending  Purchase at Taylor-Gutierrez for $615.85 on 20...   
3        approved    Travel transaction of $4852.39 at Jones-Russell   
4        approved  Purchase at

Loan to income ratio measures how large a person's loan is in comparison to their income.
if the ratio is < 1 - more income than loan-implies low risk
if the ratio is >1  - less income than loan - implies high risk
if the ratio is =1 - loan equals to income - implies moderate risk

3.Interest amount on loan

In [30]:
df['Interest Amount'] = (df['Loan Amount'] * (df['Interest Rate'] / 100)) * (df['Loan Term (Months)'] / 12)


Interest amount can be calculated in this way to understand the true cost.

In [ ]:
df

**Filtering or Aggregating Data**

1.Individuals having highest income

In [33]:
df[df['Income Level'] > 100000]

Age Occupation Risk Tolerance     Investment Goals  Income Level  \
4      36   Engineer            Low    Income Generation     106777.95   
6      70    Student           High          Speculation     120054.43   
32     52     Artist            Low               Growth     110564.54   
33     35     Lawyer           High          Speculation     104698.91   
39     75   Engineer           High               Growth     123100.16   
...   ...        ...            ...                  ...           ...   
4973   37     Doctor           High               Growth     121389.67   
4979   73    Retiree            Low          Speculation     105394.75   
4993   47    Teacher            Low    Income Generation     119478.23   
4994   30   Engineer         Medium  Wealth Preservation     130332.27   
4996   20    Student            Low  Wealth Preservation     100107.46   

                                                Address  Account Balance  \
4           96690 Campbell Common\nDwaynetown, MI 71857         77918.97   
6     997 Hernandez Parkways Apt. 451\nAmyland, MP 6...        100000.00   
32    4252 Eric Prairie Apt. 627\nGeraldhaven, KY 18564        100000.00   
33         1572 Miller Summit\nWest Hollyview, MN 18700        100000.00   
39    075 Esparza Roads Suite 226\nPort Cynthia, RI ...        100000.00   
...                                                 ...              ...   
4973  25741 Thomas Springs Apt. 592\nWest David, GA ...         38372.95   
4979  5056 Stafford Parkway Suite 412\nBrittneyberg,...        100000.00   
4993  26304 Petersen Parkways Apt. 446\nLake Davidbu...        100000.00   
4994       751 Cassidy Fords\nNorth Jamesfurt, MS 86214         48018.74   
4996  1156 Christopher Mountain Suite 011\nEast Jimm...         29879.73   

      Deposits  Withdrawals  Transfers  ...  Investments  Loan Amount  \
4     13853.25     22699.32   12189.32  ...      9725.21      50000.0   
6     23213.91     27510.89   11522.77  ...     20114.44      50000.0   
32    45803.34     19776.50   11838.40  ...     13849.15      50000.0   
33    46957.87     27563.80    8534.06  ...     22641.21      50000.0   
39    27264.37     15459.07    5789.51  ...     33202.70      50000.0   
...        ...          ...        ...  ...          ...          ...   
4973  17258.80      6539.03    3443.61  ...      2207.08      50000.0   
4979  22327.11      9932.29   16071.07  ...      9387.67      50000.0   
4993  36987.99      7848.98   12273.61  ...     12727.36      50000.0   
4994  10625.51      7343.64    6728.61  ...      7718.63      50000.0   
4996   5384.31      6262.80    3931.47  ...      2529.55      50000.0   

            Loan Purpose Employment Status Loan Term (Months)  Interest Rate  \
4       Medical Expenses          Employed                 24         10.60%   
6       Medical Expenses          Employed                 24          9.90%   
32        Small Business     Self-Employed                 48         13.38%   
33         Auto Purchase        Unemployed                 24          9.29%   
39      Medical Expenses          Employed                 60         10.91%   
...                  ...               ...                ...            ...   
4973       Auto Purchase        Unemployed                 24          9.18%   
4979    Medical Expenses     Self-Employed                 36         14.40%   
4993  Debt Consolidation          Employed                 36         10.13%   
4994      Small Business        Unemployed                 12         12.34%   
4996       Auto Purchase     Self-Employed                 24          9.38%   

     Loan Status                            Transaction Description Cash Flow  \
4       approved  Purchase at Jensen, Porter and Webb for $562.2... -23187.90   
6       approved  Transaction at Young, Martin and Henry for $15... -16775.84   
32       pending            Transaction at King-Martin for $1296.53  12169.00   
33      approved        $2803.99 spent at 

820 individuals have income higher than 100000

2.Individuals having high income and age less than 30

In [34]:
df[(df['Income Level'] > 100000) & (df['Age'] <= 30)]


Age Occupation Risk Tolerance     Investment Goals  Income Level  \
45     18     Artist           High          Speculation     128564.64   
78     21    Retiree           High          Speculation     120473.06   
84     22     Doctor         Medium               Growth     108748.71   
146    29    Student         Medium  Wealth Preservation     115933.67   
177    21     Doctor            Low               Growth     119749.28   
...   ...        ...            ...                  ...           ...   
4880   20     Artist         Medium               Growth     108853.66   
4923   25    Student           High          Speculation     101811.85   
4942   25     Doctor           High          Speculation     125088.74   
4994   30   Engineer         Medium  Wealth Preservation     130332.27   
4996   20    Student            Low  Wealth Preservation     100107.46   

                                                Address  Account Balance  \
45    5477 Rodriguez Mills Apt. 665\nEast Jennifermo...        100000.00   
78    0542 Laura Light Suite 820\nNorth Harry, MA 37325        100000.00   
84         24179 Wise Stravenue\nBurchchester, PA 73396        100000.00   
146   76124 Suarez Freeway Suite 985\nNorth Brian, N...         82546.67   
177               9748 Cynthia River\nWardton, SD 20274        100000.00   
...                                                 ...              ...   
4880  788 David Inlet Suite 396\nNew Jacobport, OK 0...        100000.00   
4923        2336 Bates Meadow\nNew Tannerfurt, MT 87157        100000.00   
4942  0684 Christopher Walk Apt. 272\nNew Maryshire,...         86834.39   
4994       751 Cassidy Fords\nNorth Jamesfurt, MS 86214         48018.74   
4996  1156 Christopher Mountain Suite 011\nEast Jimm...         29879.73   

      Deposits  Withdrawals  Transfers  ...  Investments  Loan Amount  \
45    32224.73      7813.83   11894.19  ...     24564.90     50000.00   
78    31446.97     13840.59   16313.05  ...     38084.36     50000.00   
84    25407.05     22638.97   12669.90  ...      8319.42     50000.00   
146   38451.12     23919.30   19673.66  ...     12047.67     50000.00   
177   46964.95     28237.69   18479.97  ...     15437.79     34851.81   
...        ...          ...        ...  ...          ...          ...   
4880  27700.00     16488.12   22406.50  ...      9587.76     50000.00   
4923  37245.25     18703.09   15427.01  ...     30132.45     50000.00   
4942  24940.81      6973.60    5453.84  ...     34168.79     50000.00   
4994  10625.51      7343.64    6728.61  ...      7718.63     50000.00   
4996   5384.31      6262.80    3931.47  ...      2529.55     50000.00   

            Loan Purpose Employment Status Loan Term (Months)  Interest Rate  \
45      Medical Expenses        Unemployed                 12         14.19%   
78    Debt Consolidation        Unemployed                 24         14.44%   
84      Home Improvement        Unemployed                 60         13.07%   
146     Medical Expenses           Retired                 36          9.36%   
177   Debt Consolidation        Unemployed                 60         10.35%   
...                  ...               ...                ...            ...   
4880      Small Business           Retired                 48         10.91%   
4923       Auto Purchase     Self-Employed                 24         13.69%   
4942    Home Improvement     Self-Employed                 48          9.29%   
4994      Small Business        Unemployed                 12         12.34%   
4996       Auto Purchase     Self-Employed                 24          9.38%   

     Loan Status                            Transaction Description Cash Flow  \
45      rejected         Transaction at Walker-Sanchez for $1857.15  10450.70   
78      rejected           $2691.78 spent at Gill Inc on 2024-03-16    -72.27   
84       pending             Payment of $3095.16 at Hawkins-Barnett -12601.34   
146     approved  Clothing transaction of 

3.Occupation of people having highest income

In [35]:
df[df['Income Level'] > 100000][['Income Level', 'Occupation']]


Income Level Occupation
4        106777.95   Engineer
6        120054.43    Student
32       110564.54     Artist
33       104698.91     Lawyer
39       123100.16   Engineer
...            ...        ...
4973     121389.67     Doctor
4979     105394.75    Retiree
4993     119478.23    Teacher
4994     130332.27   Engineer
4996     100107.46    Student

[820 rows x 2 columns]

Individuals from all fields score high income

4.Top 10 highest income individuals and their occupation

In [36]:
top_income = df.nlargest(10, 'Income Level')[['Income Level', 'Occupation']]
top_income

Income Level Occupation
1997     185124.37   Engineer
2040     183399.49     Lawyer
1005     176385.45    Student
1361     172857.74    Student
3246     172454.81     Doctor
1982     169498.69     Doctor
472      166915.11     Lawyer
2127     166133.81    Retiree
2387     164195.00     Lawyer
4556     159198.15    Teacher

5.Individuals having high loan

In [37]:
high_loans = df[df['Loan Amount'] >= 50000]
high_loans

Age Occupation Risk Tolerance     Investment Goals  Income Level  \
4      36   Engineer            Low    Income Generation     106777.95   
5      77     Doctor            Low    Income Generation      59157.22   
6      70    Student           High          Speculation     120054.43   
11     55    Retiree         Medium    Income Generation      79685.22   
12     50    Teacher           High  Wealth Preservation      79642.74   
...   ...        ...            ...                  ...           ...   
4988   64     Artist           High  Wealth Preservation      99420.41   
4993   47    Teacher            Low    Income Generation     119478.23   
4994   30   Engineer         Medium  Wealth Preservation     130332.27   
4995   21   Engineer           High    Income Generation      92640.59   
4996   20    Student            Low  Wealth Preservation     100107.46   

                                                Address  Account Balance  \
4           96690 Campbell Common\nDwaynetown, MI 71857         77918.97   
5     23020 Jackson View Apt. 985\nHaydenstad, MO 79409         61520.73   
6     997 Hernandez Parkways Apt. 451\nAmyland, MP 6...        100000.00   
11    5133 Pearson Hill Suite 653\nHeatherfort, CO 1...        100000.00   
12    3565 Brandon Island Suite 097\nLake Charlesbur...        100000.00   
...                                                 ...              ...   
4988              10822 Kim Courts\nKevinberg, MN 02143         56481.23   
4993  26304 Petersen Parkways Apt. 446\nLake Davidbu...        100000.00   
4994       751 Cassidy Fords\nNorth Jamesfurt, MS 86214         48018.74   
4995          43377 Brooke Gateway\nKyleville, IL 60381        100000.00   
4996  1156 Christopher Mountain Suite 011\nEast Jimm...         29879.73   

      Deposits  Withdrawals  Transfers  ...  Investments  Loan Amount  \
4     13853.25     22699.32   12189.32  ...      9725.21      50000.0   
5     22052.86     16970.19   10103.04  ...      4842.19      50000.0   
6     23213.91     27510.89   11522.77  ...     20114.44      50000.0   
11    26927.25     29938.09    8591.90  ...      6106.05      50000.0   
12    47527.94     20277.18    9533.68  ...     20108.02      50000.0   
...        ...          ...        ...  ...          ...          ...   
4988  16740.91      8955.19   11899.48  ...     11064.74      50000.0   
4993  36987.99      7848.98   12273.61  ...     12727.36      50000.0   
4994  10625.51      7343.64    6728.61  ...      7718.63      50000.0   
4995  24890.81     17829.06   21150.16  ...     32126.21      50000.0   
4996   5384.31      6262.80    3931.47  ...      2529.55      50000.0   

            Loan Purpose Employment Status Loan Term (Months)  Interest Rate  \
4       Medical Expenses          Employed                 24         10.60%   
5         Small Business           Retired                 24         12.57%   
6       Medical Expenses          Employed                 24          9.90%   
11        Small Business        Unemployed                 36         13.23%   
12        Small Business        Unemployed                 24         13.75%   
...                  ...               ...                ...            ...   
4988      Small Business           Retired                 60         12.80%   
4993  Debt Consolidation          Employed                 36         10.13%   
4994      Small Business        Unemployed                 12         12.34%   
4995    Medical Expenses        Unemployed                 24         11.41%   
4996       Auto Purchase     Self-Employed                 24          9.38%   

     Loan Status                            Transaction Description Cash Flow  \
4       approved  Purchase at Jensen, Porter and Webb for $562.2... -23187.90   
5        pending          $2275.37 debited for Mueller LLC purchase  -6204.82   
6       approved  Transaction at Young, Martin and Henry for $15... -16775.84   
11       pending  Unusual transaction at H

6.Individuals coming under High risk category

In [38]:
df[df['Risk Tolerance'].str.lower() == 'high']

Age Occupation Risk Tolerance     Investment Goals  Income Level  \
0      40     Lawyer           High  Wealth Preservation      46044.94   
6      70    Student           High          Speculation     120054.43   
12     50    Teacher           High  Wealth Preservation      79642.74   
14     69    Teacher           High               Growth      76081.96   
18     78   Engineer           High  Wealth Preservation      50479.30   
...   ...        ...            ...                  ...           ...   
4975   48     Lawyer           High          Speculation      92377.00   
4976   36     Lawyer           High               Growth      69520.01   
4988   64     Artist           High  Wealth Preservation      99420.41   
4989   21     Doctor           High  Wealth Preservation       8131.81   
4995   21   Engineer           High    Income Generation      92640.59   

                                                Address  Account Balance  \
0           7168 Moody Meadow\nHernandezshire, PW 06016         44653.26   
6     997 Hernandez Parkways Apt. 451\nAmyland, MP 6...        100000.00   
12    3565 Brandon Island Suite 097\nLake Charlesbur...        100000.00   
14            00446 Elijah Club\nBrowningland, PR 70947          8235.63   
18              452 Kim Stravenue\nWest Jared, AZ 53851         13317.40   
...                                                 ...              ...   
4975  3461 Michael Unions Suite 676\nSouth Stephanie...        100000.00   
4976          052 Megan Haven\nEast Kaylastad, IA 07207        100000.00   
4988              10822 Kim Courts\nKevinberg, MN 02143         56481.23   
4989      905 James Ridges\nSouth Jacksonland, MT 08974         14554.41   
4995          43377 Brooke Gateway\nKyleville, IL 60381        100000.00   

      Deposits  Withdrawals  Transfers  ...  Investments  Loan Amount  \
0      9156.01      9327.70    3647.92  ...      8729.30     27010.93   
6     23213.91     27510.89   11522.77  ...     20114.44     50000.00   
12    47527.94     20277.18    9533.68  ...     20108.02     50000.00   
14     2937.37      1657.03     871.39  ...      2746.74     41476.18   
18     4211.68      2346.17    2821.16  ...      2183.98     50000.00   
...        ...          ...        ...  ...          ...          ...   
4975  20803.35     28586.19   19078.13  ...     20828.75     50000.00   
4976  29589.75      9985.45   10245.69  ...     27886.73     50000.00   
4988  16740.91      8955.19   11899.48  ...     11064.74     50000.00   
4989   3054.84      3383.41    1268.13  ...      4979.50      6105.82   
4995  24890.81     17829.06   21150.16  ...     32126.21     50000.00   

            Loan Purpose Employment Status Loan Term (Months)  Interest Rate  \
0       Medical Expenses           Retired                 36         11.94%   
6       Medical Expenses          Employed                 24          9.90%   
12        Small Business        Unemployed                 24         13.75%   
14        Small Business     Self-Employed                 48         10.93%   
18        Small Business          Employed                 60         12.86%   
...                  ...               ...                ...            ...   
4975  Debt Consolidation          Employed                 24         11.06%   
4976  Debt Consolidation     Self-Employed                 36         14.11%   
4988      Small Business           Retired                 60         12.80%   
4989  Debt Consolidation          Employed                 12          6.26%   
4995    Medical Expenses        Unemployed                 24         11.41%   

     Loan Status                            Transaction Description Cash Flow  \
0        pending  Electronics transaction of $706.18 at Sanders,...  -3902.41   
6       approved  Transaction at Young, Martin and Henry for $15... -16775.84   
12      rejected  $4726.23 debited for Thomas, Weeks and Phillip...  15174.78   
14      approved   $3086.38 spent at Salaz

7.Sum of loan Amount

In [39]:
df['Loan Amount'].sum()

np.float64(189845998.13)

8. Sum of all Income

In [40]:
df['Income Level'].sum()

np.float64(353443631.54)

9.Maximum account balance

In [41]:
df['Account Balance'].max()   

np.float64(100000.0)

10.Minimum account balance

In [42]:
df['Account Balance'].min()   

np.float64(1000.0)

11.Average of deposits

In [43]:
df['Deposits'].mean()

np.float64(18524.016957999997)

12.Median value

In [ ]:
df['Account Balance'].median()

13.Minimum Interest rate

In [ ]:
df['Interest Rate'].min()

14.Maximum Interst rate

In [ ]:
df['Interest Rate'].max()

15.Individuals doing majority in jobs

In [ ]:
df['Occupation'].mode()

16.Number of jobs or unique values in Occupation

In [ ]:
df['Occupation'].nunique()

17.Loan status of Individuals

In [ ]:
df['Loan Status'].value_counts()

18. Index of Maximum value in income

In [ ]:
df['Income Level'].idxmax()

19.Index of Minimum value in income

In [ ]:
df['Income Level'].idxmin()

20.Minimum index of loan amount

In [ ]:
df['Loan Amount'].idxmin()

21.Average value, maximum value and count of Income of individuals based on their occupation

In [ ]:
df.groupby('Occupation')['Income Level'].agg(['mean', 'max', 'count'])

22.Majority people in risk category

In [ ]:
df['Risk Tolerance'].mode()

23.avg, max and min values of Income

In [ ]:
df.agg({'Income Level': ['mean', 'max', 'min']})

24.Sum , median and standard deviation of Loan Amount

In [ ]:
df.agg({'Loan Amount': ['sum', 'median', 'std']})

25.count and unique values in category, deposits

In [ ]:
df.agg({'Deposits': ['count', 'nunique']})

26.Average Loan amount per employment status

In [ ]:
df.groupby('Employment Status')['Loan Amount'].mean()

27. Average loan amount per risk tolerance and employment status

In [44]:
df.groupby(['Risk Tolerance', 'Employment Status'])['Loan Amount'].mean()

C:\Users\Irene Fathima\AppData\Local\Temp\ipykernel_2484\3645386894.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby(['Risk Tolerance', 'Employment Status'])['Loan Amount'].mean()


Risk Tolerance  Employment Status
High            Employed             37934.397936
                Retired              37051.184186
                Self-Employed        38360.315985
                Unemployed           37935.889577
Low             Employed             37596.767936
                Retired              37307.111335
                Self-Employed        38641.890605
                Unemployed           39144.265841
Medium          Employed             37887.543639
                Retired              37551.459434
                Self-Employed        37432.271505
                Unemployed           38835.328910
Name: Loan Amount, dtype: float64

28.Sum and  average deposits per income level

In [45]:
df.groupby('Income Level')['Deposits'].agg(['sum', 'mean'])

sum      mean
Income Level                    
-33320.97       394.17    394.17
-31211.06       438.30    438.30
-25488.15       299.47    299.47
-25087.86       496.90    496.90
-18262.59       351.32    351.32
...                ...       ...
 172454.81    13982.16  13982.16
 172857.74    44325.28  44325.28
 176385.45    35944.08  35944.08
 183399.49    23570.93  23570.93
 185124.37     4846.21   4846.21

[5000 rows x 2 columns]

 Individuals of each occupation falling under loan status.



In [46]:
df.pivot_table(index='Occupation', columns='Loan Status', aggfunc='count', values='Income Level')

C:\Users\Irene Fathima\AppData\Local\Temp\ipykernel_2484\1493857534.py:1: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  df.pivot_table(index='Occupation', columns='Loan Status', aggfunc='count', values='Income Level')


Loan Status  approved  pending  rejected
Occupation                              
Artist            312      200       190
Doctor            330      198       195
Engineer          361      190       211
Lawyer            299      187       181
Retiree           289      220       207
Student           327      205       195
Teacher           323      200       180

 Average loan amounts for each Occupation and Risk Tolerance level.

In [47]:
df.pivot_table(index='Occupation', columns='Risk Tolerance', values='Loan Amount')

C:\Users\Irene Fathima\AppData\Local\Temp\ipykernel_2484\108689399.py:1: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  df.pivot_table(index='Occupation', columns='Risk Tolerance', values='Loan Amount')


Risk Tolerance          High           Low        Medium
Occupation                                              
Artist          38236.021944  39455.209286  37612.141637
Doctor          36781.804826  38222.775098  37196.747773
Engineer        38373.423574  39282.627862  38683.597806
Lawyer          38029.817336  38645.955971  37764.394101
Retiree         36699.685485  37555.193931  39100.277880
Student         37855.255342  37841.655236  37197.916192
Teacher         38571.425652  36180.929205  38141.600342

**Converting into Proper case**

In [48]:
df['Loan Status'] = df['Loan Status'].str.title()
df

Age Occupation Risk Tolerance     Investment Goals  Income Level  \
0      40     Lawyer           High  Wealth Preservation      46044.94   
1      30    Teacher            Low  Wealth Preservation      57169.50   
2      37    Teacher            Low          Speculation      71760.86   
3      27    Student         Medium          Speculation     -25488.15   
4      36   Engineer            Low    Income Generation     106777.95   
...   ...        ...            ...                  ...           ...   
4995   21   Engineer           High    Income Generation      92640.59   
4996   20    Student            Low  Wealth Preservation     100107.46   
4997   53     Lawyer            Low          Speculation       6058.18   
4998   75    Student         Medium               Growth      76214.25   
4999   33    Retiree            Low               Growth      62638.63   

                                                Address  Account Balance  \
0           7168 Moody Meadow\nHernandezshire, PW 06016         44653.26   
1          50001 Hector Square\nWest Luisfurt, MA 51935         29175.47   
2           997 James Isle\nNorth Rebeccafurt, RI 13366         86141.59   
3      1607 Joshua Camp Apt. 634\nConleymouth, CT 66479          1000.00   
4           96690 Campbell Common\nDwaynetown, MI 71857         77918.97   
...                                                 ...              ...   
4995          43377 Brooke Gateway\nKyleville, IL 60381        100000.00   
4996  1156 Christopher Mountain Suite 011\nEast Jimm...         29879.73   
4997  92818 Mary Garden Suite 343\nGuerramouth, DE 6...          9673.78   
4998  6197 Stephanie Circles Apt. 765\nCherylstad, N...         90780.00   
4999                   Unit 4064 Box 6179\nDPO AA 30544         62218.05   

      Deposits  Withdrawals  Transfers  ...  Investments  Loan Amount  \
0      9156.01      9327.70    3647.92  ...      8729.30     27010.93   
1      5933.22      8671.60    6729.86  ...      4545.18     31266.97   
2     22583.11     16468.58    6032.53  ...      8251.45     41260.58   
3       299.47       289.09     109.83  ...       195.73      5000.00   
4     13853.25     22699.32   12189.32  ...      9725.21     50000.00   
...        ...          ...        ...  ...          ...          ...   
4995  24890.81     17829.06   21150.16  ...     32126.21     50000.00   
4996   5384.31      6262.80    3931.47  ...      2529.55     50000.00   
4997   3274.37      1832.16    1742.00  ...       898.41      7054.49   
4998  35679.58     12022.55   12832.11  ...      6370.88     34847.26   
4999  22903.01     16866.44    4295.50  ...      3886.95     12459.69   

          Loan Purpose Employment Status Loan Term (Months)  Interest Rate  \
0     Medical Expenses           Retired                 36         11.94%   
1        Auto Purchase           Retired                 36          8.08%   
2        Auto Purchase          Employed                 12         13.07%   
3       Small Business           Retired                 60          6.73%   
4     Medical Expenses          Employed                 24         10.60%   
...                ...               ...                ...            ...   
4995  Medical Expenses        Unemployed                 24         11.41%   
4996     Auto Purchase     Self-Employed                 24          9.38%   
4997  Medical Expenses     Self-Employed                 60         10.02%   
4998  Home Improvement           Retired                 48          8.49%   
4999  Medical Expenses     Self-Employed                 12          7.95%   

     Loan Status                            Transaction Description Cash Flow  \
0        Pending  Electronics transaction of $706.18 at Sanders,...  -3902.41   
1       Approved            Transaction at Evans-Smith for $2250.03 -10287.93   
2        Pending  Purchase at Taylor-Gutierrez for $615.85 on 20...  -1444.33   
3       Approved    Travel transaction of $4852.39 at Jones-Russel

**Filtering based on Occupation**

In [49]:
df_engineer = df[df['Occupation'].str.contains('engineer', case=False, na=False)]
df_engineer

Age Occupation Risk Tolerance     Investment Goals  Income Level  \
4      36   Engineer            Low    Income Generation     106777.95   
18     78   Engineer           High  Wealth Preservation      50479.30   
22     60   Engineer         Medium               Growth      47649.79   
28     77   Engineer            Low    Income Generation      56181.95   
30     61   Engineer           High  Wealth Preservation      88941.51   
...   ...        ...            ...                  ...           ...   
4935   74   Engineer         Medium          Speculation      53889.95   
4977   85   Engineer         Medium          Speculation      66148.32   
4982   46   Engineer         Medium  Wealth Preservation      86255.89   
4994   30   Engineer         Medium  Wealth Preservation     130332.27   
4995   21   Engineer           High    Income Generation      92640.59   

                                                Address  Account Balance  \
4           96690 Campbell Common\nDwaynetown, MI 71857         77918.97   
18              452 Kim Stravenue\nWest Jared, AZ 53851         13317.40   
22    840 Sanders Orchard Apt. 159\nGloriaborough, V...         94400.83   
28                           USCGC Miller\nFPO AE 24947         91997.80   
30            15488 Michael Way\nWest Michael, SD 98778        100000.00   
...                                                 ...              ...   
4935      59571 Angela Turnpike\nPhillipsside, SC 04351         32070.81   
4977         0913 Crystal Valleys\nLake Kevin, AL 65584        100000.00   
4982  9315 Contreras Junction Suite 406\nMarkborough...        100000.00   
4994       751 Cassidy Fords\nNorth Jamesfurt, MS 86214         48018.74   
4995          43377 Brooke Gateway\nKyleville, IL 60381        100000.00   

      Deposits  Withdrawals  Transfers  ...  Investments  Loan Amount  \
4     13853.25     22699.32   12189.32  ...      9725.21     50000.00   
18     4211.68      2346.17    2821.16  ...      2183.98     50000.00   
22    27493.93     20428.22   10403.07  ...     17067.74      6087.60   
28    13196.44     10630.77   10449.70  ...      6409.27     36721.87   
30    12025.25     21662.70    6638.52  ...      5026.06     50000.00   
...        ...          ...        ...  ...          ...          ...   
4935  13859.93      4383.76    6292.24  ...      3909.87     50000.00   
4977  42791.73     10127.00    7306.52  ...     19630.08     50000.00   
4982  26062.25     19628.09    5691.39  ...     18198.12     50000.00   
4994  10625.51      7343.64    6728.61  ...      7718.63     50000.00   
4995  24890.81     17829.06   21150.16  ...     32126.21     50000.00   

            Loan Purpose Employment Status Loan Term (Months)  Interest Rate  \
4       Medical Expenses          Employed                 24         10.60%   
18        Small Business          Employed                 60         12.86%   
22    Debt Consolidation          Employed                 60          8.02%   
28      Home Improvement        Unemployed                 48          8.34%   
30    Debt Consolidation           Retired                 60          9.29%   
...                  ...               ...                ...            ...   
4935    Home Improvement          Employed                 48         11.72%   
4977  Debt Consolidation          Employed                 12         11.83%   
4982       Auto Purchase     Self-Employed                 36          9.09%   
4994      Small Business        Unemployed                 12         12.34%   
4995    Medical Expenses        Unemployed                 24         11.41%   

     Loan Status                            Transaction Description Cash Flow  \
4       Approved  Purchase at Jensen, Porter and Webb for $562.2... -23187.90   
18      Rejected             Payment of $2437.30 at Whitehead-Ochoa   -958.96   
22      Rejected  $2379.82 spent at Harris, Russell and Herrera ...  -5107.35   
28      Approved            $333.99 debite

In [50]:
df['Income Level'].var()

np.float64(882907027.6664587)

In [51]:
df['Loan Amount'].std()

np.float64(15085.14656191004)